src/main/flight/pid.c

```diff
@@ -538,6 +567,79 @@ static void handleItermRotation()
     }
 }
 
+#ifdef USE_ACRO_TRAINER
+
+int8_t acroTrainerSign(float x)
+{
+    return (x > 0.0f) - (x < 0.0f);
+}
+
+// Acro Trainer - Manipulate the setPoint to limit axis angle while in acro mode
+// There are three states:
+// 1. Current angle has exceeded limit
+//    Apply correction to return to limit (similar to pidLevel)
+// 2. Future overflow has been projected based on current angle and gyro rate
+//    Manage the setPoint to control the gyro rate as the actual angle  approaches the limit (try to prevent overshoot)
+// 3. If no potential overflow is detected, then return the original setPoint
+
+static float applyAcroTrainer(int axis, const rollAndPitchTrims_t *angleTrim, float setPoint)
+{
+    float ret = setPoint;
+
+#ifndef SIMULATOR_BUILD
+    if (!FLIGHT_MODE(ANGLE_MODE) && !FLIGHT_MODE(HORIZON_MODE)) {
+        bool resetIterm = false;
+        float projectedAngle;
+        const int8_t setpointSign = acroTrainerSign(setPoint);
+        const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
+        const int8_t angleSign = acroTrainerSign(currentAngle);
+
+        // Limit and correct the angle when it exceeds the limit
+        if ((fabsf(currentAngle) > acro_trainer_angle_limit) || (acroTrainerAxisState[axis] != 0)) {
+            if (angleSign == -setpointSign) {  // stick has reversed - stop limiting
+                acroTrainerAxisState[axis] = 0;
+            } else {
+                if (acroTrainerAxisState[axis] != angleSign) {
+                    acroTrainerAxisState[axis] = angleSign;
+                    resetIterm = true;
+                }
+                ret = ((acro_trainer_angle_limit * angleSign) - currentAngle) * ACRO_TRAINER_GAIN;
+            }
+        }
+        
+        // Not currently over the limit so project the angle based on current angle and
+        // gyro angular rate using a sliding window based on gyro rate (faster rotation means larger window.
+        // If the projected angle exceeds the limit then apply limiting to minimize overshoot.
+        if (acroTrainerAxisState[axis] == 0) {
+            const timeUs_t checkInterval = MAX(lrintf(constrainf(fabsf(gyro.gyroADCf[axis]) / 500.0f, 0.0f, 1.0f) * acro_trainer_lookahead_time), ACRO_TRAINER_LOOKAHEAD_MIN);
+            projectedAngle = (gyro.gyroADCf[axis] * checkInterval * 0.000001f) + currentAngle;
```

Comment:
Basically just experimentally came up with this. I just found that at any rate faster than 500dps it worked best to use the entire look-ahead window and scale the window to be smaller at lower rates.  When using too large a window at low rate it introduced an over-correction.  On the other hand using too small a window didn't introduce enough correction soon enough at high gyro rates.  So this is what I experimentally came up with.  I tried many (last count was 8 I think) other approaches that used much more complicated logic but ultimately the simple approach worked as well or better.  Also put it in context that we're not looking for awesome flight behavior with somebody slamming against the limit.  It's more designed to let them "lean" against it while learning (primarily LOS) acro flight.

Reply:
This approach is problematic .
It may work only on your configuration.
And nobody in future will understand it, probably 'improving' it to suit his configuration/needs. The code will rot beyond repair (it took ages to remove multiwii heritage - strange heuristics and 8bit optimizations)

Ideally some simplified physical model would be used as starting condition  and control strategy derived from it ...